Task feature dev

rrich360 opened the pull request:
I made a few changes to the task 1 assignment and finally managed to get the asynchronous calls working in angularjs. added  jetty plugin to pom.

Commit:
adding task 1 files for review/help

Commit:
update from last pull request

Commit:
update to pull request for task 1

rrich360 committed:
Delete target directory

rrich360 committed:
Delete src/main/java/org/custom/task1 directory

rrich360 committed:
Update to pom.xml

The pom.xml from the initial pull request was still included to this new pull request so I deleted the unnecessary portion of the pom.xml in order to ensure the new files recently pushed to the repo will work on the jetty server properly.

Mbd06b commented on .classpath:
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
.gitignore

Mbd06b commented on .project:
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
+ .gitignore

Mbd06b commented on .settings/.jsdtscope:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
+ .gitignore

Mbd06b commented on .settings/org.eclipse.core.resources.prefs:
```diff
@@ -0,0 +1,5 @@
+eclipse.preferences.version=1
```
+ .gitignore

Mbd06b commented on .settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,21 @@
+eclipse.preferences.version=1
```
+ .gitignore

Mbd06b commented on .settings/org.eclipse.m2e.core.prefs:
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
```
+ .gitignore

Mbd06b commented on .settings/org.eclipse.wst.common.component:
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?><project-modules id="moduleCoreId" project-version="1.5.0">
```
+ .gitignore

Mbd06b commented on .settings/org.eclipse.wst.common.project.facet.core - Copy.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
+ .gitignore

Mbd06b commented on .settings/org.eclipse.wst.common.project.facet.core.xml:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
+ .gitignore

Mbd06b commented on .settings/org.eclipse.wst.jsdt.ui.superType.container:
```diff
@@ -0,0 +1 @@
+org.eclipse.wst.jsdt.launching.baseBrowserLibrary
```
+ .gitignore

Mbd06b commented on .settings/org.eclipse.wst.jsdt.ui.superType.name:
```diff
@@ -0,0 +1 @@
+Window
```
+ .gitignore

Mbd06b commented on .settings/org.eclipse.wst.validation.prefs:
```diff
@@ -0,0 +1,2 @@
+disabled=06target
```
+ .gitignore

Mbd06b commented on pom.xml:
```diff
@@ -0,0 +1,90 @@
+			<version>1.2</version>
+		</dependency>
+		<dependency>
+			<groupId>org.slf4j</groupId>
+			<artifactId>slf4j-api</artifactId>
+			<version>1.7.30</version>
```
slf4j-api is found inside slf4j-simple delete this unnecessary dependency.

Mbd06b commented on pom.xml:
```diff
@@ -0,0 +1,90 @@
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-validation</artifactId>
```
If this isn't used at this point, remove it.

Mbd06b commented on src/main/java/com/rogerr/custom/controller/HomePageController.java:
```diff
@@ -0,0 +1,19 @@
+			@GetMapping("/header")
+		public String showHomePage() {
+			
+			System.out.println("In Home Page");
```
use the slf4j logger

Mbd06b commented on src/main/java/com/rogerr/custom/controller/LoginController.java:
```diff
@@ -0,0 +1,41 @@
+		Login user=new Login();
+		model.addAttribute("Login",user);
+		
+		return"user_login";
```
use camelCase as a reference for the jsp(s)/views

Mbd06b commented on src/main/java/com/rogerr/custom/controller/ProductController.java:
```diff
@@ -0,0 +1,40 @@
+		return "product_all";
+	}
+	
+	@GetMapping("/{productId}")
+	public String getProductById(Model model, @PathVariable("productId")String productId) {
+		
+		model.addAttribute("product",productService.getProductById(productId));
+		return "product";
+		
+	}
```
the mappings suggest that you're building a restul endpoint. However, you are returning a view, consider returning a "productDetailPage" view here, to clarify the difference between a product (page) and product (resource)

Mbd06b commented on src/main/java/com/rogerr/custom/controller/SubscriberController.java:
```diff
@@ -0,0 +1,16 @@
+
+	  @RequestMapping(method = RequestMethod.GET)
+	    public String getSubscribersPage() {
+	        return "UserManagement";
```
camelCase references to jsp/views

Mbd06b commented on src/main/java/com/rogerr/custom/controller/UserController.java:
```diff
@@ -0,0 +1,120 @@
+    @Autowired
+    private UserService userService;
+
+    private static Logger logger = LoggerFactory.getLogger(UserController.class);
```
yay! using the logger. Nooice!

Mbd06b commented on src/main/java/com/rogerr/custom/controller/UserController.java:
```diff
@@ -0,0 +1,120 @@
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.setLocation(ucBuilder.path("/user/{id}").buildAndExpand(user.getId()).toUri());
+        return new ResponseEntity<Void>(headers, HttpStatus.CREATED);
```
install an IDE-based linter like Sonarlint or FindBugs.  I don't think specifying void type is necessary, just <>

Mbd06b commented on src/main/java/com/rogerr/custom/controller/UserController.java:
```diff
@@ -0,0 +1,120 @@
+    public ResponseEntity<?> updateUser(@PathVariable("id") long id, @RequestBody User user) {
+        logger.debug("Updating User " + id);
+
+        User currentUser = userService.findById(id);
+
+        if (currentUser == null) {
+            logger.debug("User with id " + id + " not found");
+            return new ResponseEntity<User>(HttpStatus.NO_CONTENT);
+        }
+
+        currentUser.setUsername(user.getUsername());
+        currentUser.setAddress(user.getAddress());
+        currentUser.setEmail(user.getEmail());
```
I can understand this might make sense if you haven't added persistence yet, however, after task 2, your objects should have a reference to their own id (if they have been persisted before), so you can just pass the whole object (with it's updates) to the service and most repository implementations are smart enough to handle the update for you.

Mbd06b commented on src/main/java/com/rogerr/custom/controller/UserController.java:
```diff
@@ -0,0 +1,120 @@
+    // ------------------- Delete All Users--
+
+    @RequestMapping(value = "/user/", method = RequestMethod.DELETE)
+    public ResponseEntity<?> deleteAllUsers() {
```
lol, let's do it.

Mbd06b commented on src/main/java/com/rogerr/custom/model/Login.java:
```diff
@@ -0,0 +1,30 @@
+public class Login {
+	
+	
+	@NotBlank(message="username can NOT be empty!")
```
ATM I would avoid using Java Validation on your objects, can cause A LOT of hang-ups/slowdowns when you're getting started stick to field-level validations in the views themselves.  If it's just this, that's fine, but be careful.

Mbd06b commented on src/main/java/com/rogerr/custom/model/Product.java:
```diff
@@ -0,0 +1,51 @@
+	private double price;
+	private String storeOwner;
+	
+	public Product(String id, String name, String description, double price, String storeOwner) {
+		super();
+		this.id = id;
+		this.name = name;
+		this.description = description;
+		this.price = price;
+		this.storeOwner = storeOwner;
+	}
```
Avoid custom constructors, unless you really need them. Also IF you define a custom constructor, you need to define the default or constructor, particularly when you attempt to use these POJOS for persistence later.

Mbd06b commented on src/main/java/com/rogerr/custom/model/Product.java:
```diff
@@ -0,0 +1,51 @@
+	private String id;
+	private String name;
+	private String description;
+	private double price;
```
mostly avoid primitives in your Models because they can't be null.   Also, instead of doubles use BigDecimal in properties used for monetary values.

Mbd06b commented on src/main/java/com/rogerr/custom/model/User.java:
```diff
@@ -0,0 +1,83 @@
+		this.email = email;
+	}
+
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + (int) (id ^ (id >>> 32));
+		return result;
+	}
+
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (!(obj instanceof User))
+			return false;
+		User other = (User) obj;
+		if (id != other.id)
+			return false;
+		return true;
+	}
```
don't overload the hashCode or equals() methods unless you have a specific reason to do so.

Mbd06b commented on src/main/java/com/rogerr/custom/model/User.java:
```diff
@@ -0,0 +1,83 @@
+
+	@Override
+	public String toString() {
+		return "User {id=" + id + ", username=" + username + ", address=" + address
+				+ ", email=" + email + "}";
```
don't put PII in your toString methods.

Mbd06b commented on src/main/webapp/WEB-INF/views/UserManagement.jsp:
```diff
@@ -0,0 +1,122 @@
+<%@ include file="header.jsp" %>
+
+<%@ page language="java" contentType="text/html; charset=ISO-8859-1" pageEncoding="ISO-8859-1"%>
```
this jsp page declaration should be 1st line on page, also, change default charset in your IDE for jsp to UTF-8

Mbd06b commented on src/main/webapp/WEB-INF/views/UserManagement.jsp:
```diff
@@ -0,0 +1,122 @@
+<%@ include file="header.jsp" %>
+
+<%@ page language="java" contentType="text/html; charset=ISO-8859-1" pageEncoding="ISO-8859-1"%>
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core"%>
```
taglib declarations should go above your "includes"

Mbd06b commented on src/main/webapp/WEB-INF/views/UserManagement.jsp:
```diff
@@ -0,0 +1,122 @@
+                          <div class="form-group col-md-12">
+                              <label class="col-md-2 control-lable" for="file">Email</label>
+                              <div class="col-md-7">
+                                  <input type="email" data-ng-model="ctrl.user.email" name="email" class="email form-control input-sm" placeholder="Enter your Email" required/>
```
declare a unique id= attribute for each of your inputs.

Mbd06b commented on src/main/webapp/WEB-INF/views/UserManagement.jsp:
```diff
@@ -0,0 +1,122 @@
+                              <td><span data-ng-bind="u.address"></span></td>
+                              <td><span data-ng-bind="u.email"></span></td>
+                              <td>
+                              <button type="button" data-ng-click="ctrl.edit(u.id)" class="btn btn-success custom-width">Edit</button>  
+                              <button type="button" data-ng-click="ctrl.remove(u.id)" class="btn btn-danger custom-width">Remove</button>
```
add  a unique id attribute for every interactive element

Mbd06b commented on src/main/webapp/WEB-INF/views/UserManagement.jsp:
```diff
@@ -0,0 +1,122 @@
+
+                      <div class="row">
+                          <div class="form-group col-md-12">
+                              <label class="col-md-2 control-lable" for="file">Email</label>
```
for="file" here I don't think is resolved anywhere.  In HTML, the label for= attribute, is paired with an id= attribute, usually of the input being labeled.

Mbd06b commented on src/main/webapp/WEB-INF/views/header.jsp:
```diff
@@ -0,0 +1,43 @@
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core"%>
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <title>Gizmos World</title>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1">
```
this is good meta here, specifying language, charset, and viewport.

Mbd06b commented on src/main/webapp/WEB-INF/views/product_all.jsp:
```diff
@@ -0,0 +1,22 @@
+		<p>Web Store Owner :  ${product.storeOwner} </p>  
+        </div>
+        <div class="panel-footer"> <button type="button" class="btn btn-primary btn-md"
+           onClick="location.href='/products/${product.id} ' ">Review</button></div>
```
interactive elements need a unique id= attribute.

Mbd06b commented on src/main/webapp/WEB-INF/views/product_all.jsp:
```diff
@@ -0,0 +1,22 @@
+    </div>
+    </c:forEach>
+  </div>
+</div><br>
```
avoid line break tags where possible.  also, use the self-closing syntax `<br />`

rrich360 replied:
Updated

Mbd06b commented on src/main/webapp/WEB-INF/views/user_login.jsp:
```diff
@@ -0,0 +1,79 @@
+</head>
+
+ 
+<link href="//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css" rel="stylesheet">
```
link should be in the `<head>` element.

rrich360 replied:
Updated

Mbd06b commented on src/main/webapp/WEB-INF/views/user_login.jsp:
```diff
@@ -0,0 +1,79 @@
+
+ 
+<link href="//maxcdn.bootstrapcdn.com/font-awesome/4.2.0/css/font-awesome.min.css" rel="stylesheet">
+
+
+<div class="container">
```
missing `<body>` tag, and you have `<div>` and other content being declared outside semantic html tags, These are known as semantic landmarks and are important for 508 compliance. https://www.w3schools.com/html/html5_semantic_elements.asp

rrich360 replied:
By the way, thanks for providing the link for this. It's been updated

Mbd06b commented on src/main/webapp/WEB-INF/views/user_login.jsp:
```diff
@@ -0,0 +1,79 @@
+<!DOCTYPE html>
+<html>
```
specify default lang on the html tag

rrich360 replied:
Updated

Mbd06b commented on src/main/webapp/static/js/service/user_service.js:
```diff
@@ -0,0 +1,38 @@
+
+function UserServiceFactory($http) {
+
+    var REST_SERVICE_URI = 'http://localhost:8080/user/';
```
don't hardcode your protocol//host:port

Mbd06b reviewed (changes requested):
looking good, I would definitely  run your webpages through  https://validator.w3.org/
 and use the Axe devtools as a good exercise in clean accessible html.

chriswbluer commented on src/main/webapp/customerDataAngularJS.html:
```diff
@@ -0,0 +1,216 @@
+    src="https://ajax.googleapis.com/ajax/libs/angularjs/1.4.4/angular.js"></script>
+ 
+<title>Spring Boot AngularJS example</title>
+<script type="text/javascript">
```
Could this be moved to it's own separate file and be imported? For consistency? 

For a reference, I do it here: https://github.com/chriswbluer/Pokedex/blob/main/webapp/src/main/webapp/WEB-INF/views/pokedexManagement.jsp#L155

rrich360 replied:
Not sure why this file was even added as I did not push this with the new 42 files for this request. I think this file file is from my older PR. My b, I thought I removed this.

rrich360 committed:
Delete customerDataAngularJS.html

This file was not included with new 42 files that were pushed for second PR. I made some changes to the original code from the first PR, but this file was not part of it, therefore, I am deleting it.

rrich360 committed:
Delete index.jsp

This file is not part of the newer 42 files pushed via the second, most recent PR. I tried pulling the file then pushing the replacement files, but index.jsp still appears. Therefore, I am deleting this file directly from this repo for clarity.

rrich360 committed:
Delete MANIFEST.MF

This file is an older file and is not part of the new 42 files pushed for the second PR.

rrich360 commented:
I added new generic user to login. When the user logs in there is a welcome page. The welcome page view is named "success.jsp". This is being pushed to Task 2.
